src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs

```diff
@@ -140,53 +142,68 @@ public partial class NetworkChange
         private static unsafe void CreateSocket()
         {
             Debug.Assert(s_socket == null, "s_socket is not null, must close existing socket before opening another.");
-            int newSocket;
+            IntPtr newSocket;
             Interop.Error result = Interop.Sys.CreateNetworkChangeListenerSocket(&newSocket);
             if (result != Interop.Error.SUCCESS)
             {
                 string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
                 throw new NetworkInformationException(message);
             }
 
-            s_socket = new SocketWrapper(newSocket);
+            s_socket = new Socket(new SafeSocketHandle(newSocket, ownsHandle: true));
 
-            new Thread(args => LoopReadSocket((SocketWrapper)args!))
-            {
-                IsBackground = true,
-                Name = ".NET Network Address Change"
-            }.UnsafeStart(s_socket);
+            // Don't capture ExecutionContext.
+            ThreadPool.UnsafeQueueUserWorkItem(
+                static socket => ReadEventsAsync(socket),
+                s_socket, preferLocal: false);
         }
 
         private static void CloseSocket()
         {
             Debug.Assert(s_socket != null, "s_socket was null when CloseSocket was called.");
-            Interop.Error result = Interop.Sys.CloseNetworkChangeListenerSocket(s_socket != null ? s_socket.Socket : -1);
-            if (result != Interop.Error.SUCCESS)
-            {
-                string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
-                throw new NetworkInformationException(message);
-            }
-
+            s_socket.Dispose();
             s_socket = null;
         }
 
-        private static unsafe void LoopReadSocket(SocketWrapper initiallyCreatedSocket)
+        private static async void ReadEventsAsync(Socket socket)
         {
-            while (s_socket == initiallyCreatedSocket) //if both references are equal then s_socket was not changed
+            try
             {
-                //we can continue processing events
-                Interop.Sys.ReadEvents(initiallyCreatedSocket.Socket, &ProcessEvent);
+                while (true)
+                {
+                    // Wait for data to become available.
+                    await socket.ReceiveAsync(Array.Empty<byte>(), SocketFlags.None).ConfigureAwait(false);
+
+                    Interop.Error result = ReadEvents(socket);
+
+                    if (result != Interop.Error.SUCCESS ||
+                        result != Interop.Error.EAGAIN)
+                    {
+                        return;
+                    }
+                }
             }
+            catch (ObjectDisposedException)
+            { } // Socket disposed.
+            catch (SocketException se) when (se.SocketErrorCode == SocketError.OperationAborted)
+            { } // ReceiveAsync aborted by disposing Socket.
+            catch
+            { } // Unexpected error.
```

Comment:
`ReadEvents` now makes the error available.
I also split up the expected exceptions.

There is no logging (yet).

Reply:
@MihaZupan, what should be done here around logging?  Or should we crash for an unexpected error?

Reply:
Would such an error previously be swallowed by `SystemNative_ReadEvents`?
If that is the case, I would prefer we don't crash the process outright, though I don't see any harm in adding a `Debug.Fail` + `NetEventSource.Log` if it does occur.

Reply:
@MihaZupan I've added the logging, ptal.

Reply:
Thanks, that looks good to me